content/struct-classes.md

```diff
@@ -0,0 +1,502 @@
+specially. A `val` prefix is implicitly added to such a parameter, unless the parameter
+already carries a `val` or `var` modifier. Hence, an accessor definition for the
+parameter is generated.
+
+A struct class definition `c[tps](ps_1)...(ps_n)` with type parameters `tps` and
+value parameters `ps` is handled as follows. A `private` method named `copy` is
+implicitly added to the class definition unless the class already has a member
+(directly defined or inherited) with that name, or the class has a repeated parameter.
+The method is defined as follows:
+
+~~~ scala
+private def copy[Ts](ps_1')...(ps_n'): c[Ts] = c[Ts](xs_1)...(xs_n)
+~~~
+
+Where `Ts` is the vector of types defined in the type parameter section of the class
+definition. Each `xs_i` denotes the parameter names of the parameter section `ps_i`.
+The value parameters `ps_1j'` of the first parameter list have the form
+`x_1j: T_1j = this.x_1j`. The other parameters `ps_ij'` of the `copy` method are
+defined as `x_ij: T_ij`. In all cases, `x_ij` and `T_ij` refer to the name and type
+of the corresponding class parameter `ps_ij`.
+
+Every struct class implicitly overrides some method definitions of class `scala.AnyRef`
+unless a definition of the same method is already given in the struct class itself or
+a concrete definition of the same method is given in some base class of the struct
+class different from `AnyRef`. In particular:
+
+
+- Method `def equals(that: Any): Boolean` is structural equality, where two instances
+  are equal if they both belong to the struct class in question and they have equal
+  (with respect to `equals`) constructor arguments (restricted to the class’s elements,
+  i.e., the first parameter list)
+- Method `def hashCode(): Int` computes a hash-code. If the `hashCode` method of the
+  data structure members map equal (with respect to `equals`) values to equal hash-codes,
+  then the struct class `hashCode` method does too
+- Method `def toString(): String` returns a string representation which contains the
+  name of the class and its elements.
+
+Finally, mirrors (instances of `scala.deriving.Mirror`) are not synthesized by the
+compiler for struct classes.
+
+### Compatibility
+
+The changes described in this proposal are backward binary compatible because they do
+not affect the bytecode produced by the existing language features.
+
+The changes described in this proposal are backward TASTy compatible because they
+don’t require any changes at the TASTy level.
+
+Source compatibility is preserved, and the semantics of existing valid programs is not
+changed because the proposal does not change the existing language features, it only
+adds a new keyword, `struct`.
+
+## Alternatives
+
+Besides the two solutions shown in the Motivation section (based on regular classes
+or case classes), we also considered a more powerful variant of `struct class` that
+would also automatically generate the transformation methods (`withName`, `withAge`, 
```

Comment:
I feel not having this kind of feature will be frustrating to users. Most will reach out to struct classes as a way to deal with lighter-weight case classes or solve the bin compact issues, but the problem of updating and dealing with immutable structures does not go away.

We have a similar problem in Kotlin in the context of value classes where they may include special syntax for immutable copy updates https://github.com/Kotlin/KEEP/blob/master/notes/value-classes.md#updating-immutable-classes. I don't propose we follow a similar approach; just pointing out that not having a feature like this and forcing users into considerable boilerplate to get it working may introduce more issues than `struct` classes solve. In contrast, case classes that at least give you `copy` are not great to work with and with bin compact issues, but they are still an alternative to writing a bunch of boilerplate.

I think `struct` classes will be even better if they improve on `case classes` not just from a bin compact standpoint but also the state of what working with immutable data looks like in Scala.